### ISSUE TITLE
Chat messages disappearing until pull torefresh or app kill

### ISSUE THREAD

**Issue by jagas** [chat-sdk/chat-sdk-android#163]
Chat messages in any thread first appears when you send it, but if you go back and go back into chat that message is not there.

**Comment by brucexia** [comment]
Noticed same issue

**Comment by bensmiley** [comment]
I can't replicate this. Just to confirm:
1. Open 1-to-1 thread
2. Send text message
3. Message appears
4. Click back
5. Re-open thread

After doing this, I see the message as expected. If you are seeing something different, can you give me exact steps to reproduce the issue?

**Comment by jagas** [comment]
Yes, I can confirm that these are exact steps I take to reproduce it. 
I also noticed t hat after step 3 if you wait some time (not long, lets say approx 30 sec or a 60sec) it will stay when you come back. 

Maybe this will shed some light on that:
when I write a message and go back I also see that I have new message from myself in that chat and I can see it as "lastMessage" under the username, but when I go in I cannot see it unless I wait sometime. Some kind of sync happens

**Comment by bensmiley** [comment]
I can't replicate this issue. Does it happen on default version of the ChatSDK from Github with no modification? If so, can you give me details of the hardware you're using and the Android device?

**Comment by jagas** [comment]
I did compile new fresh version from Github with one slight modification, but its styling issue (app crashing when inflating view)
problem still happens every time I send a message, it happens in every chat, always.

Android version 6.0.1 hardware: samsung note 4

**Comment by bensmiley** [comment]
Can you attach a video?

**Comment by jagas** [comment]
Hi,

here is a video showing that when I write a message in group or 1-to-1 chat they do not appear there. this is not modified version except the tabs style thing, so it would compile

**Comment by bensmiley** [comment]
I can't reproduce this. Does it happen with the completely non-modified version?

**Comment by jagas** [comment]
yes, it also happens when i install demo apk that you built

**Comment by jagas** [comment]
and to add more, it happens everytime for every message. not that it happens sometimes. in other words I have not been avle to send a message properly yet

**Comment by bensmiley** [comment]
And it happens when you checkout the project and compile? Are there any exceptions? Have you tried using the latest version? There have been updates recently.

**Comment by jagas** [comment]
hi, #188 explains why it happens, my clock was one minute off and thats why it was happening. Another question is: why it still shows me a new message is in the chat when I wrote it myself?

**Comment by simonsmiley64** [comment]
The Chat is designed to mark messages from yourself as read immediately.

Is it showing the message when you send a message and quickly leave the chat or when you send a message and it updates correctly? It has been known to occur in the first instance but shouldn't with the second

**Comment by jagas** [comment]
For me it happens each time.
1. I write a message
2. wait around for 5-10 sec, press back and it shows me as unread
3. go to a different chat room, do the same thing, press back
4. now the old "unread" is gone, but new message took over and is unread

**Comment by Yariko-chan** [comment]
I'm facing the same issue.
And also after full reinstall unread messages counter displays all messages from chat (mine and not mine) as unread until I open chat. For that counter I use folowing snippet:

        ```
Thread currentThread = StorageManager.shared().fetchThreadWithEntityID(mCurrentChatId);
        int unreadCount = 0;
        if (currentThread != null) {
            unreadCount = currentThread.getUnreadMessagesCount();
        }
```

**Comment by OferRnz** [comment]
I'd like to contribute from my long experience with Google App Engine and Datastore.
After talking to the App Engine developers, these were the conclusions:

The app engine server is not a single server. It is many connected machines. The "server timestamp" may vary by machine. Note that the variation may be dependent on the user's location, which may explain why it always shows for e.g. @jagas, @Yariko-chan  but not for you.

The solution:
Always read a little bit into the past. The Google app server engineers suggested that 12 seconds should be safe. Make sure you deal with duplicate and with "past" messages correctly on the client. 

Just my 2 cents...  :)

**Comment by bensmiley** [comment]
@OferRnz Thanks for your comment. I double checked the code and I don't think that the timestamp should cause a problem. What happens in the code is that when the app starts, a listener is added which listens for any messages that have arrived since the last message. If there were an issue with the timestamp, we would expect the app to work correctly after 12 had passed (even if it didn't handle messages that had been sent within 12 seconds of the last message).

I did manage to replicate an issue which is similar to this. To replicate the issue:

1. Open the chat view
2. Put the app in the background
3. Send a message to the app from a different device
4. Re-open the app. The message will be absent from the list. 

This has now been fixed and will be available in the next Gradle release - 4.0.9. For the time being I'm closing this issue because I've not been able to replicate it. If it continues, please reopen the issue.
